Fixing UVCDAT runtime environment

remram44 opened the pull request:
Sourcing the setup_environment.sh before using the software is error-prone. Furthermore, it is not actually needed anymore when running 'uvcdat', since the wrapper script will source it; this causes a warning to be displayed (`UVCDAT setup already sourced for this install location.`).

Also, once `setup_runtime.sh` is sourced, a whole bunch of executables appear in the path, replacing the system's default (things like `python`, `curl`, `uuid`... 260 in total on my system). For the same reasons, installing UV-CDAT globally is not a good idea (will break system's `python`).

This pull request intends to make the setup_runtime logic internal, and exposes `uvcdat` and `uvcdatpython` wrappers that can safely be added to the PATH (since they are on a directory of their own).

TODO:
* [X] Change setup_runtime.sh
  * Re-running the script for same installation path is fine
  * However running for a different path fails
  * Remove shebang, should only be sourced
* [X] Create wrappers
  * Scripts that set the environment then run their conterpart
  * 'cdat' (python), 'uvcdat' (vistrails), 'activate_uvcdat' (setup_runtime.sh')
  * They get the path from @-variables, not their location; they can be installed out of the install tree (e.g. somewhere on the user's PATH)
  * [ ] uvcdat.mac.in wasn't updated, I don't know what that is
* [X] Get CMake to configure and install wrappers
  * [X] They get copied at build time or with `make wrappers` (if you need to do it as root)
  * [X] Location configured by `WRAPPER_INSTALL_LOCATION` cache variable, defaults to `${CMAKE_INSTALL_PREFIX}/wrappers` (not so useful but you can manually add on PATH)
  * `make install` cannot be used, it is broken because of its special use in CPack
* [ ] Review and test
  * I'd definitely like some feedback on this
* [ ] Update setup_runtime.csh
  * I don't know much about csh, need help

aashish24 commented:
@doutriaux1 I have to review this branch. I am wondering if you reviewed it already? 

- Aashish

remram44 commented:
This is definitely not ready, I'll try and do one more pass on this asap.

aashish24 commented:
Great.  It would be nice if you can layout a roadmap here (as serveral people made suggestions after your initial commit). Something as simple as bulleted list, nothing too fancy.

doutriaux1 commented:
as far as I can tell the major issue with it was that it fixes ONLY .sh we cannot bring it in until .csh is fixed as well (or at least confirm the issue does not exists for .csh)

durack1 commented:
Would it be possible to add csh tests to duplicate any tests that run on the sh versions? It seems kinda silly to have this environment test omitted when it's an important initialization script..

remram44 commented:
I think I'm done, apart from the CSH script. I updated the todo-list.

remram44 commented:
I have no idea what `make install` does currently, apart from installing my wrappers. Here it made a "UVCDAT" directory besides "install", that looks to be a copy, yet doesn't work if "install" is removed; can anyone explain?

painter1 commented:
At least for the diagnostics, "make install" makes two copies, one
    in a local build directory and the other under the Python's
    site-packages directory.  The one which really matters is under
    site-packages.  I think that a UV-CDAT build is similar, but more
    complicated.
    - Jeff
    On 1/21/15 3:03 PM, Remi Rampin wrote:
    
    
      
      I have no idea what make install does currently,
        apart from installing my wrappers. Here it made a "UVCDAT"
        directory besides "install", that looks to be a copy, yet
        doesn't work if "install" is removed; can anyone explain?
      —
        Reply to this email directly or view
          it on GitHub.

remram44 commented:
I think UV-CDAT's build just installs everything during the `make` phase and the `make install` is completely broken, but I'd like confirmation before I work around this issue.

doutriaux1 commented:
@remram44 you're right the make install is completely ignored at the moment. "make" does everything, I "think" it's because some packages needs other to be installed in place before build starts, @aashish24 we could probably revise that? But that would be a lot of work for not much benefit (and it would add one line to the install instructions) but at least it would be consistent with regular build systems.

remram44 commented:
If we want to keep this behavior it's fine but "make install" should display an error message instead. I can move the wrapper-creation code back to the "make" step.

doutriaux1 commented:
you're right "make install" should probably simply output a message in the lines of
```
UV-CDAT is already installed at: ${CMAKE_INSTALL_PREFIX}
```

aashish24 commented:
make install is not that useful in this case as we have most of the packages installed externally and its much easier to get them installed during the first step verus in two steps. 

Said that, we can have empty / no operation make instal and display the message that @doutriaux1 suggested.

remram44 commented:
> Said that, we can have empty / no operation make instal and display the message that @doutriaux1 suggested.

Yes, it seems that CMake doesn't generate a "make install" target unless the INSTALL() function is used, so it should only be a matter of finding and removing these.

remram44 commented:
[804abe8](https://github.com/UV-CDAT/uvcdat/blob/804abe81adbdc7e11437501fa046ff7cd2bb7e55/CMakeLists.txt#L747-L868) is the culprit.

aashish24 commented:
@remram44 we cannot get rid of that as we will need that for the DMG (and RPM in the future). That install rule is executed during the CPACK time. what is the exact issue you are having? Could you plesea remind me again?

remram44 commented:
The problem is that "make install" exists and does nothing sane (it copies everything to a new "UVCDAT" directory from which binaries don't even run (they are built for `.../install`)). "make install" was the perfect place for the wrapper installation step; this is what every UNIX user expects, and it is usually run as root.
Can't you change the target name (to "dmg-install" or something) or take it out since it is not meant for users?

To avoid permission issues, I could make a "make installwrappers" or something that puts "cdat" and "uvcdat" in /usr/local/bin (actually WRAPPER_BINARIES_LOCATION).

aashish24 commented:
In that case, I would suggest you generate the wrapper during make time. I would not encourage us to do two step (make and make install) since make in our case does the installation for you. I will check on make install rule bit more before I do much. The problem is that install is expected for cpack. I don't know how it work other way round.

aashish24 commented:
@remram44 what you can do is you can make your wrapper dependent on the CDAT in that case it will be run last. Also, you want want to make it a external module (unless there is a strong reason not do so).

remram44 commented:
Installing the wrappers to system directories will require root permissions, that's why the install step exists in the first place.

aashish24 commented:
we should't be installing to the system directories. We install things where the CMAKE_INSTALL_PREFIX is set. Please use that.

remram44 commented:
Here it is again: the point of this branch is to have "cdat" and "uvcdat" wrappers, together with "setup_runtime", in separate directories that can be safely exposed to the user without bringing in all the UV-CDAT environment which would definitely break the user's.
If I'm putting these in install/bin this effort has no point whatsoever.

aashish24 commented:
@remi one thing you could do is define a new variable WRAPPER_INSTALL_PREFIX (which can defualt to user's home directory or something reasonable). If you do that, and define this as a external target, then someone can just do this: 

make wrapper (call it something else )

to me, this makes much more sense. Also, some users might have not root permissions (often in a control environment). So installing it at /usr should n't be the default IMO. Let me know what you think of it. Thanks

remram44 commented:
The path is not exactly a prefix, since I don't append eg. `/bin` at the end. The variable is the actual directory into which the wrappers are put. It is currently named `WRAPPER_BINARIES_LOCATION`.
The users only need root permissions if that variable is set to a location they can't write to (which is not the default). They can also copy/move them anywhere, they won't break. But it's tedious IMHO if there is a make target that only does that.
No other option I suppose, if CPack cannot be tamed. I'll do that soon.

There is a possibility of having the `activate_uvcdat` script load in a `deactivate` command, the same way Python's virtualenv system work. Preliminary work (for bash only, I'm afraid I still don't know anything about csh) is [here](https://github.com/remram44/uvcdat/commit/2749393f60715c35af3ed1d01ec68349dba5142c) but it needs unit tests. Is that of interest?

aashish24 commented:
- WRAPPER_BINARIES_LOCATION sounds good to me. You can also think of more conventional name 

WRAPPER_INSTALL_LOCATION

- Right, the root permission issue is independent of whether or not we install it during the make or make install time

- I am less willing to change the CPack stuff (that is already complicated)

- activate_uvcdat and deactivate would be nice!! I would have something short though (loadcdat unloadcdat)

remram44 commented:
Alright, I rebased everything. Changes:
* `WRAPPER_BINARIES_LOCATION` -> `WRAPPER_INSTALL_LOCATION`
* Wrappers installed by the `wrappers` target (e.g. `make wrappers`), run by default (e.g. `make` or `make all`)
  * They can still be copied from install/wrappers/ to somewhere else if you didn't change `WRAPPER_INSTALL_LOCATION`, they will run from anywhere
* Renamed `activate_uvcdat` to `loadcdat`

aashish24 commented:
@remram44 thanks. Much appreciated. I will review it and get it merged today.

remram44 commented:
Note (cf roadmap) that uvcdat.mac.in and setup_runtime.csh were not updated.

aashish24 commented on CMake/cdat_modules_extra/setup_runtime.sh.in:
```diff
@@ -39,6 +35,10 @@ library_paths=( @SETUP_LIBRARY_PATHS@ )
 python_paths=( @SETUP_PYTHON_PATHS@ )
 executable_paths=( @SETUP_EXECUTABLE_PATHS@ )
 
+if [ -z "$UVCDAT_DISABLE_PROMPT" ] ; then
```
we should document it somewhere...

aashish24 commented:
LGTM :+1: except the documentation part.

durack1 commented:
@aashish24 @doutriaux1 just wondering whether this should be closed when the *.csh variant hasn't been updated?

doutriaux1 commented:
@aashish24 I push a PR to "undo" the prompt change. I understand the idea, but it should be done right i.e
* give which version of uvcdat you're using
* maybe last part of the directory of install
* make it prettier than this

In the mean time I turn this ON only if requested, having like that by default is just not right it think

doutriaux1 commented:
also why did we approve this before csh is done. We need to keep these both in sync @durack1 and @jypeter are not the only csh users out there. We can't just blow them away because we don't like tcsh (nothing worng with not liking tcsh but we have to support it anyway :wink: )

aashish24 commented:
The wrappers won't break the existing means to run uvcdat right? As far as we can don't break the current functionality I think it's okay to add new one. Thoughts? We won't annouce the feature until then. Also if remi can push a fix for csh soon then also this should be fine.

remram44 commented:
Oh, I didn't intend to update the CSH version myself... If a CSH user could do it, that would be a lot faster (and probably safer) than if I have to learn about CSH first.

doutriaux1 commented:
@durack1 or @jypeter it would be awesome if you could help on this end. I could do it but it's going to take my weeks/months before I can get to it.
